File: CP/LeetCode/Python/20.py
# 20. Valid Parentheses

from __future__ import annotations


class Solution:
    def isValid(self, s: str) -> bool:
        # Time: O(n)
        # Space: O(n)
        # stack = []
        # for i in s:
        #     if i in "({[":
        #         stack.append(i)
        #     else:
        #         if not stack:
        #             return False
        #         if i == ")" and stack[-1] != "(":
        #             return False
        #         if i == "}" and stack[-1] != "{":
        #             return False
        #         if i == "]" and stack[-1] != "[":
        #             return False
        #         stack.pop()
        # return not stack

        while "()" in s or "{}" in s or "[]" in s:
            s = s.replace("()", "").replace("{}", "").replace("[]", "")
        return not s


if __name__ == "__main__":
    s = Solution()
    print(s.isValid("()"))  # True
    print(s.isValid("()[]{}"))  # True
    print(s.isValid("(]"))  # False
    print(s.isValid("([)]"))  # False

File: CP/LeetCode/C/20.c
// 20. Valid Parentheses

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct Node
{
    char data;
    struct Node *next;
};

struct Node *TOP = NULL;

int push(char element)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));
    new_node->data = element;
    new_node->next = TOP;
    TOP = new_node;

    return 0;
}

char pop()
{
    struct Node *temp = TOP;
    TOP = TOP->next;
    char popped = temp->data;
    free(temp);

    return popped;
}

bool isValid(char *s)
{
    int i = 0;
    char ch;

    while (s[i] != '\0')
    {
        if (s[i] == '(' || s[i] == '{' || s[i] == '[')
        {
            push(s[i]);
        }
        else
        {
            if (TOP == NULL)
            {
                return false;
            }

            ch = pop();
            if (s[i] == ')' && ch != '(')
            {
                return false;
            }

            if (s[i] == '}' && ch != '{')
            {
                return false;
            }

            if (s[i] == ']' && ch != '[')
            {
                return false;
            }
        }
        i++;
    }

    if (TOP == NULL)
    {
        return true;
    }

    return false;
}

int main()
{
    char s[] = "()";
    int valid = isValid(s);
    printf("%d\n", valid);

    return 0;
}